spark-tsne-core/src/main/scala/com/github/saurfang/spark/tsne/TSNEGradient.scala

```diff
@@ -20,7 +23,7 @@ object TSNEGradient extends Logging  {
     val num: DenseMatrix[Double] = (y1(::, *) + sumY).t // k * n
     num := 1.0 :/ (1.0 :+ (num(::, *) + sumY(idx).toDenseVector)) // k * n
 
-    idx.indices.foreach(i => num.unsafeUpdate(i, idx(i), 0.0)) // num(i, i) = 0
+    idx.indices.foreach(i => num.update(i, idx(i), 0.0)) // num(i, i) = 0
```

Comment:
Was `unsafeUpdate` replaced by `update` in the library? I remembered I used `unsafeUpdate` because it doesn't do boundary check. Or have you benchmarked the differences and found they are insignificant?